### DISCUSSION TITLE
Adding highway:platform as parameter in lua profile?

### DISCUSSION THREAD

**Discussion by PhilipNgo** [valhalla/valhalla#3565] [Q&A] [answered]
Hi!
I was wondering if there is a way to include "highway:platform" for the pedestrian lua profile.
Is there a reason that it is not included since I think it would be great for navigating pedestrians from one bus/train platform to another?

**Comment by nilsnolde** [comment]
IMO that's mostly interesting when you're serving multi-modal transit routing no?

if you're building with public transit, that info is part of the GTFS datasets valhalla ingests (currently served by transit.land, soon hopefully raw GTFS files). it's better data than in osm I'd say, since the transit providers are QC'ing the locations etc.

but you would like to have platform ways available for normal routing, outside of transit routing? why route to platforms if you wouldn't be able to use public transit? maybe I'm misunderstanding smth :sweat_smile:

**Reply by kevinkreiser** [reply]
this was my take as well. we deliberately excluded it because we get it from GTFS -> Transitland and if we are not importing transit data then its not actually useful. the problem would be deduplicating anything that came with transit data vs anything that comes in osm. we decided to just get it all from one place rather than to worry about that deduplication. its true that adding transit is optional but having the stations/platforms is not as useful if you cant get on/off busses and trains

**Comment by slvlirnoff** [comment]
It could be useful for pedestrians routing through train stations / complex pedestrian bus stops, depending on how the station has been mapped, e.g. for instance for platform "areas".

**Comment by PhilipNgo** [comment]
It might be me that's misunderstanding the potential of what can be done with Valhalla 😅. It is true that it is mostly interesting with multi-modal transit routing.  The intent was initially that we currently have starts and stops and that Valhalla would be used to guide the pedestrians through complex stations as @slvlirnoff mentioned.

Is it possible to use Valhalla for travel with respect to bus/train arrival/departure time?

**Reply by nilsnolde** [reply]
> Is it possible to use Valhalla for travel with respect to bus/train arrival/departure time?

yep, it's supporting transit (and traffic-influenced) time-based routing, see e.g. the docs https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md

we also soon will have support (or have already? didn't really update myself for a few weeks) for in-door routing, which you might find useful as well. of course train stations will have to be appropriately tagged for that to work. and then you'd still have to get the platform locations from GTFS data. not sure how to deal with your problem where you want to use it outside of real transit routing.

**Comment by slvlirnoff** [comment]
Even without multimodal, i.e. walking through/next a train/bus station, if no walking path explicitly put on platform. Also in my experience transit stops are defined somewhere in a platform areas and often far from any pedestrian way.

I don't know about the status of walking through areas in Valhalla right now, is it possible as a beeline? Or does it follows the edges of the area? If the latter then probably best not to include because it will create ridiculously long pedestrian legs.

**Reply by nilsnolde** [reply]
open-space routing per se is not supported. there's multiple models for that though, been published over the past years, and I always had that on my wish list for implementation (also helps a lot for navigating pedestrians on big squares which are mapped as areas with no ways in the square). it's a semi-big project though and you know, time etc.. problem is also that it'd likely be a pre-processing step where we add "virtual" edges leading through open space, bloating the graph somewhat, but we could make an opt-in config option I'm sure.

so yeah, valhalla does follow the edges of any multipolygon relation if they're accessible, it doesn't route through open areas (yet).

**Reply by kevinkreiser** [reply]
small correction here, i was pretty sure we dont route on any area (polygon) features: https://github.com/valhalla/valhalla/blob/master/lua/graph.lua#L1399-L1402

and i think we only use relations to get metadata like route numbers (and in a separate processing administrative bounaries)

**Reply by nilsnolde** [reply]
Ah yeah right, not explicitly as relations.. didn’t word it very well I just realized.. often/usually the ways defining such a multipolygon are routable ways, then we indirectly use them, guess I meant that:)

**Reply by PhilipNgo** [reply]
If one is restricted to only navigating from one stop to another (given stops and start from other service) and not using the transit functionality of Valhalla, is it reasonable to create a custom cost function/lua profile that handles highway:platforms and by adding custom tags in osm to specify the pathway of pedestrians to handle open area paths. Or is it just better to add the custom tags and let Valhalla handle the rest?

If the former is it reasonable to follow this https://github.com/valhalla/valhalla/pull/1756/files and add a new custom profile?

**Reply by PhilipNgo** [reply]
Not sure if I am clear in my writing haha. But essentially, I just want to navigate from one bus/train stop to another.

**Comment by kevinkreiser** [comment, marked as answer]
@PhilipNgo we could add a flag to the mjilnor config that optionally keeps those ways. We would just ad the parsing inthe Lua and use the flag in c++ to determine whether or not to keep them once they come in from the Lua parser. We probably wouldn't have to do more than that

**Reply by PhilipNgo** [reply]
That would be very appreciated! So that I understand correctly, it would be when building Valhalla, we would just trigger the flag and the pedestrian profile would be able to walk on the platforms as well?

**Reply by kevinkreiser** [reply]
We dont have "profiles". The data we create has all attributes and all modes of access. We would:

1. add a boolean flag to the mjolnir section of the config, call it `include_platforms`
2. add lua parsing to enable these ways with pedestrian access, set their use to an existing transit platform one or create a new one
3. read the flag on the c++ side and if its disabled and we see a transit platform or whatever use come through from lua we skip it

Still though I'm not sure if this is going to do what you want. I dont know the state of highway=platform and how it is being used at present. Some example I have seen are tiny little island disconnected from everything else in which case they are fairly useless.

**Reply by PhilipNgo** [reply]
I think that sounds good and would probably work for my use case. As of now, the highway=platform is the thing blocking the pedestrian from reaching the correct destination.

Here is kind of what I am after:
![wanted](https://user-images.githubusercontent.com/34343658/157707293-1f967fa5-1b86-4b88-a846-ea8de454977c.png)
 

And here is what is currently happening due to the platform:
Here the behaviour is kind of what we want, it can find stairs etc as pedestrians as expected.
![stairs](https://user-images.githubusercontent.com/34343658/157707668-ec9c2888-73f9-4ce4-b67f-30acf93ffb33.png)

Here is where it takes a detour because of the platform(thingy circled in blue):
![D1](https://user-images.githubusercontent.com/34343658/157707896-fbec0f44-7b4f-4ac3-acd6-502f88af6184.jpg)

Where I will be using platform wont be disconnected to any pedestrians paths (at least i am pretty sure)

**Reply by slvlirnoff** [reply]
I feel that you might not get exactly that result. Since platforms are generally mapped as 'area', most of the time the composite ways aren't always explicitly connected to nearby stairs or pedestrian ways (that are inside the area polygon, because it's implicit) so you might not be able to route to / from the platform ways since valhalla doesn't create that vertex for you.

Also in some cases depending on the connection between the platforms ways and the nearby pedestrian ways and the 'edge' candidates of your search, you might end up walking all around the platform in a much more dramatic effect than now.

**Reply by kevinkreiser** [reply]
@slvlirnoff yeah exactly! this is what i was refering to when i was saying i didnt check to see how it was mapped.

long term what we should do if we want to route on area features is compute for polygons what is called a skeleton or medial axis and use that as the edges to traverse the interior of the area feature. This would work well for long skinny features which happen to be basically edges in the graph but are mapped as polygons, but for larger more convex features it would create some interesting problems to overcome. it would then also need to connect this medial axis/skeleton to the existing road/pedestiran network much the same way we do when adding transit data to the graph. i can write up a proposal on this but its quite a large effort, though it feels like some really fun code to write wrt algorithms :smile:

**Reply by PhilipNgo** [reply]
Then maybe the easiest way to do it is maybe to add more tags in OSM/ custom OSM map that Valhalla takes in so that it has the correct routes? Or do you guys have any suggestions on how you would solve this problem? 😅
